Detectors/MUON/MCH/Workflow/src/DataDecoder.cxx

```diff
@@ -0,0 +1,602 @@
+  {
+    return buffers[previousBufId];
+  }
+
+  int getCurrentBufId()
+  {
+    return currentBufId;
+  }
+
+  int getPreviousBufId()
+  {
+    return previousBufId;
+  }
+
+  void mergeDigits();
+};
+
+//_________________________________________________________________
+class MergerBase
+{
+ public:
+  MergerBase() = default;
+  ~MergerBase() = default;
+
+  virtual void setDigitHandler(std::function<void(const Digit&)> h) = 0;
+  virtual void setOrbit(int feeId, uint32_t orbit, bool stop) = 0;
+  virtual void addDigit(int feeId, int solarId, int dsAddr, int chAddr,
+                        int deId, int padId, int adc, Digit::Time time, Digit::Time stopTime) = 0;
+  virtual void mergeDigits(int feeId) = 0;
+};
+
+class MergerSimple : public MergerBase
+{
+  std::function<void(const Digit&)> sendDigit;
+
+ public:
+  MergerSimple() = default;
+  ~MergerSimple() = default;
+
+  void setOrbit(int feeId, uint32_t orbit, bool stop)
+  {
+  }
+
+  void setDigitHandler(std::function<void(const Digit&)> h)
+  {
+    sendDigit = h;
+  }
+
+  void addDigit(int feeId, int solarId, int dsAddr, int chAddr,
+                int deId, int padId, int adc, Digit::Time time, Digit::Time stopTime)
+  {
+    sendDigit(o2::mch::Digit(deId, padId, adc, time));
+  }
+
+  void mergeDigits(int feeId) {}
+};
+
+//_________________________________________________________________
+class Merger : public MergerBase
+{
+  int32_t feeId = {-1};
+  FeeIdMerger mergers[MCH_MERGER_FEEID_MAX + 1];
+
+ public:
+  Merger() = default;
+  ~Merger() = default;
+
+  void setOrbit(int feeId, uint32_t orbit, bool stop);
+
+  void setDigitHandler(std::function<void(const Digit&)> h);
+
+  void addDigit(int feeId, int solarId, int dsAddr, int chAddr,
+                int deId, int padId, int adc, Digit::Time time, Digit::Time stopTime);
+
+  void mergeDigits(int feeId);
+};
+
+void FeeIdMerger::setOrbit(uint32_t orbit, bool stop)
+{
+  // send digits of previous orbit if either the stop RDH is received
+  // or a new orbit is started
+  if ((orbit == buffers[currentBufId].orbit) && (!stop)) {
+    return;
+  }
+
+  int nSent = 0;
+  for (auto& d : buffers[previousBufId].digits) {
+    if (!d.merged && (d.digit.getPadID() >= 0)) {
+      uint16_t size = (d.stopTime.getBXTime() - d.digit.getTime().getBXTime()) / 4 + 1;
+      sendDigit(d.digit);
+      nSent += 1;
+    }
+  }
+  if (mPrint) {
+    std::cout << "[FeeIdMerger] sent " << nSent << " digits for orbit " << buffers[previousBufId].orbit << "  current orbit is " << orbit << std::endl;
+  }
+
+  currentBufId = 1 - currentBufId;
+  previousBufId = 1 - previousBufId;
+  buffers[currentBufId].digits.clear();
+  buffers[currentBufId].orbit = orbit;
+}
+
+void FeeIdMerger::mergeDigits()
+{
+  auto checkDigits = [](MergerDigit& d1, MergerDigit& d2) -> bool {
+    // skip digits that are already merged
+    if (d1.merged || d2.merged)
+      return false;
+
+    // skip all digits not matching the detId/padId
```

Comment:
I would externalise this check to e.g. a free function (bool areSamePad(const MergedDigit& d1, const MergedDigit& d2) for instance) or a comparison operator of the MergedDigit

Reply:
Done.